common/djangoapps/util/model_utils.py

```diff
@@ -0,0 +1,146 @@
+            tracking
+
+    Returns:
+        dict: a mapping of field names to current database values of those
+            fields, or an empty dicit if the model is new
+    """
+    try:
+        old_model = model_class.objects.get(pk=instance.pk)
+    except model_class.DoesNotExist:
+        # Object is new, so fields haven't technically changed.  We'll return
+        # an empty dict as a default value.
+        return {}
+    else:
+        field_names = [
+            field[0].name for field in model_class._meta.get_fields_with_model()
+        ]
+        changed_fields = {
+            field_name: getattr(old_model, field_name) for field_name in field_names
+            if getattr(old_model, field_name) != getattr(instance, field_name)
+        }
+
+        return changed_fields
+
+
+def emit_field_changed_events(instance, user, event_name, db_table, excluded_fields=None, hidden_fields=None):
+    """Emits an event for each field that has changed.
+
+    Note that this function expects that a `_changed_fields` dict has been set
+    as an attribute on `instance` (see `get_changed_fields_dict`.
+
+    Args:
+        instance (Model instance): the model instance that is being saved
+        user (User): the user that this instance is associated with
+        event_name (str): the name of the event to be emitted
+        db_table (str): the name of the table that we're modifying
+        excluded_fields (list): a list of field names for which events should
+            not be emitted
+        hidden_fields (list): a list of field names specifying fields whose
+            values should not be included in the event (None will be used
+            instead)
+
+    Returns:
+        None
+    """
+    def clean_field(field_name, value):
+        """
+        Prepare a field to be emitted in a JSON serializable format.  If
+        `field_name` is a hidden field, return None.
+        """
+        if field_name in hidden_fields:
+            return None
+        # Country is not JSON serializable.  Return the country code.
+        if isinstance(value, Country):
+            if value.code:
+                return value.code
+            else:
+                return None
+        return value
+
+    excluded_fields = excluded_fields or []
+    hidden_fields = hidden_fields or []
+    changed_fields = getattr(instance, '_changed_fields', {})
+    for field_name in changed_fields:
+        if field_name not in excluded_fields:
+            old_value = clean_field(field_name, changed_fields[field_name])
+            new_value = clean_field(field_name, getattr(instance, field_name))
+            emit_setting_changed_event(user, event_name, db_table, field_name, old_value, new_value)
+    # Remove the now inaccurate _changed_fields attribute.
+    if hasattr(instance, '_changed_fields'):
+        del instance._changed_fields
+
+
+def emit_setting_changed_event(user, event_name, db_table, setting_name, old_value, new_value):
```

Comment:
I think we should just have this method always use the event_name "edx.users.settings.changed" instead of passing it into the method. We always use the same event name, and as it is, the event name has to be specified a bunch of places. This method is specifically about eventing a change in setting.